--- a/cpu/65816/cpu_65c816.h
+++ b/cpu/65816/cpu_65c816.h
@@ -53,6 +53,7 @@
 		void *context{};
 		void (*idle)(void *, uint32_t){};
 		bool (*last_cycle)(void *, uint8_t){};
+		void (*halt)(void *, bool){};
 		uint32_t (*read)(void *, uint32_t, uint8_t *){};
 		uint32_t (*write)(void *, uint32_t, uint8_t){};
 	} timed_bus;
--- a/cpu/65816/cpu_65c816.cc
+++ b/cpu/65816/cpu_65c816.cc
@@ -372,9 +372,11 @@
 
 void WDC65C816::WAI()
 {
+	if(timed_bus.halt) timed_bus.halt(timed_bus.context, false);
 }
 void WDC65C816::STP()
 {
+	if(timed_bus.halt) timed_bus.halt(timed_bus.context, true);
 }
 
 void WDC65C816::DoInterrupt(InterruptType type)
--- a/cpu/65816/cpu_65c816_instructions.inl
+++ b/cpu/65816/cpu_65c816_instructions.inl
@@ -3027,7 +3027,7 @@
 };
 struct STP
 {
-	static constexpr uint8_t opcode = 0xCB;
+	static constexpr uint8_t opcode = 0xDB;
 
 	static constexpr size_t kBytes = 1;
 	static constexpr const char kMnemonic[] = "STP";
